org-code-javabuilder/lib/src/main/java/org/code/javabuilder/UserProjectFileManager.java

```diff
@@ -1,15 +1,65 @@
 package org.code.javabuilder;
 
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.*;
+
+/** Manages the in-memory list of user project files. */
 public class UserProjectFileManager implements ProjectFileManager {
+  private final String baseUrl;
+  private final String[] fileNames;
+  private final ArrayList<ProjectFile> fileList;
 
-  public UserProjectFileManager(String baseUrl, String[] fileNames) {}
+  public UserProjectFileManager(String baseUrl, String[] fileNames) {
+    this.baseUrl = baseUrl;
+    this.fileNames = fileNames;
+    this.fileList = new ArrayList<>();
+  }
 
+  /**
+   * Asynchronously loads all user project files.
+   *
+   * @throws UserFacingException If the load times out or fails.
+   */
   public void loadFiles() throws UserFacingException {
-    throw new UserFacingException("We couldn't load your files");
+    ExecutorService executor = Executors.newCachedThreadPool();
+    List<Callable<Boolean>> callables = new ArrayList<>();
+    for (String fileName : fileNames) {
+      ProjectFile file = new ProjectFile(fileName);
+      fileList.add(file);
+      callables.add(new ProjectFileLoader(file, String.join("/", baseUrl, fileName)));
+    }
+    List<Future<Boolean>> futures;
+    try {
+      futures = executor.invokeAll(callables);
+      executor.shutdown();
+      boolean finished;
+      finished = executor.awaitTermination(10, TimeUnit.SECONDS);
+
+      if (!finished) {
+        throw new UserFacingException(
```

Comment:
I kind of expected UserFacingException to only be thrown for errors that are caused by the students code, but now I see that it's more for errors we want to report to the user.  I wonder if it would be useful to add another exception class to separate errors related to student code and reportable errors related to our code.

Reply:
✔️